cabal-install/src/Distribution/Client/CmdRepl.hs

```diff
@@ -274,29 +292,35 @@ replAction flags@NixStyleFlags { extraFlags = (replFlags, envFlags), ..} targetS
           replFlags' = case originalComponent of
             Just oci -> generateReplFlags includeTransitive elaboratedPlan' oci
             Nothing  -> []
-          replFlags'' = case replType of
-            GlobalRepl scriptPath
-              | Just version <- compilerCompatVersion GHC compiler
-              , version >= minGhciScriptVersion -> ("-ghci-script" ++ scriptPath) : replFlags'
-            _                                   -> replFlags'
+
+        -- The '-ghci-script' flag is path to a GHCi script responsible for changing to the
+        -- correct directory. Only works on GHC >= 7.6, though. 🙁
+        replFlags'' <- case replType of
+          GlobalRepl | compilerCompatVersion GHC compiler >= Just minGhciScriptVersion -> do
+            let ghciScriptPath = projectRoot </> "setcwd.ghci"
+            cwd <- getCurrentDirectory
+            writeFile ghciScriptPath (":cd " ++ cwd)
+            return $ ("-ghci-script" ++ ghciScriptPath) : replFlags'
+          _ -> return replFlags'
 
         return (buildCtx, replFlags'')
 
+    incDir <- maybeToList <$> mapM (canonicalizePath . dropFileName) mScript
+    script <- if replOptionsNoLoad replFlags == Flag True
+              then return []
+              else maybeToList <$> mapM (\s -> if isAbsolute s then return s else makeRelativeToCwd s) mScript
     let buildCtx' = buildCtx
           { elaboratedShared = (elaboratedShared buildCtx)
             { pkgConfigReplOptions = replFlags
-              { replOptionsFlags = (replOptionsFlags replFlags) ++ replFlags''
+              { replOptionsFlags  = replOptionsFlags replFlags ++ replFlags'' ++ fmap ("-i" ++) incDir ++ script
```

Comment:
magic include seems weird and breaks some abstractions in my opinion. I.e. do we anywhere just randomly add include directories?

Why does the repl need it? Needs at least documentation why we add it here, I am fine with refactoring it later though.